packages/user/DiffAdjust/service/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+                initial_difficulty,
+                window_seconds,
+                target_min,
+                target_max,
+                floor_difficulty,
+                last_updated,
+                percent_change,
+            );
+            new_instance.save();
+
+            new_instance
+        }
+
+        fn percent(&self) -> f64 {
+            self.percent_change as f64 / ONE_MILLION as f64
+        }
+
+        pub fn check_difficulty_decrease(&mut self) -> u64 {
+            let now = TransactSvc::call().currentBlock().time.seconds();
+            let seconds_elapsed = (now.seconds - self.last_update.seconds) as u32;
+            let windows_elapsed = seconds_elapsed / self.window_seconds;
+            if windows_elapsed > 0 {
+                let below_target = self.counter < self.target_min;
+                let seconds_remainder = seconds_elapsed % self.window_seconds;
+                self.counter = 0;
+                self.last_update = TimePointSec::from(now.seconds - seconds_remainder as i64);
+                if below_target {
+                    let mut new_difficulty = self.active_difficulty as f64;
+                    let factor = 1.0 - self.percent();
+                    for _ in 0..windows_elapsed {
+                        new_difficulty = new_difficulty * factor;
+                    }
+                    self.active_difficulty = (new_difficulty as u64).max(self.floor_difficulty);
+                }
+            }
+            self.active_difficulty
+        }
+
+        fn check_difficulty_increase(&mut self) -> u64 {
+            if self.counter > self.target_max {
+                let percent = 1.0 + self.percent();
+                let times_over_target = self.counter / self.target_max;
+                let mut difficulty = self.active_difficulty as f64;
+                for _ in 0..times_over_target {
+                    difficulty = difficulty * percent;
+                }
+                self.active_difficulty = difficulty.min(u64::MAX as f64) as u64;
+                self.counter = 0;
+                self.last_update = TransactSvc::call().currentBlock().time.seconds();
+                self.save();
+            }
+            self.active_difficulty
+        }
+
+        fn check_sender_has_nft(&self) {
+            check(
+                Nft::call().getNft(self.nft_id).owner == get_sender(),
+                "must be owner of rate limiter",
+            );
+        }
+
+        fn check_sender_is_consumer(&self) {
+            check(self.consumer == get_sender(), "must be consumer");
+        }
+
+        pub fn increment(&mut self, increment_amount: u32) -> u64 {
+            self.check_sender_is_consumer();
+            let difficulty = self.check_difficulty_decrease();
+            self.counter += increment_amount;
+            self.check_difficulty_increase();
+            self.save();
+            difficulty
+        }
+
+        pub fn update_window(&mut self, seconds: u32) {
```

Comment:
None of the parameter update functions call `check_difficulty_[increase|decrease]`, which means that they can retroactively alter increases and decreases that were previously reported by the query.

Reply:
We're now checking for price decreases on the updates, price increase isn't necessary given it runs during a price increase

Reply:
Suppose that `target_max = 1000`. If I change it to `target_max = 10`, it could trigger up to 100 difficulty increases. Is that the desired behavior? Considering that `target_max` was `1000` at the time all those increments happened, it seems a bit counterintuitive. Also, exponential increase means that large changes like this are liable to overflow. I think it would be better if decreasing the `target_max` results in at most one immediate difficulty increase.

Reply:
I've added clamping `fn check_difficulty_increase(&mut self, clamp_increase: bool) -> u64 {`